// react-explode/Corregidor.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Corregidor;

var _react = _interopRequireWildcard(require("react"));

var _gsap = _interopRequireWildcard(require("gsap"));

var _Circle = _interopRequireDefault(require("./Icons/Circle"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var LINES_RADIUS = 47.4;
var CIRCLE_STROKE_WIDTH = 0.1;
var LINE_STROKE_WIDTH = 1;
var COUNT = 32;
var TIME_LINE = null;

function Corregidor(_ref) {
  var size = _ref.size,
      delay = _ref.delay,
      repeatDelay = _ref.repeatDelay,
      repeat = _ref.repeat,
      style = _ref.style,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? "#fff" : _ref$color,
      onComplete = _ref.onComplete,
      onStart = _ref.onStart,
      onRepeat = _ref.onRepeat,
      className = _ref.className;
  var circleRef = (0, _react.useRef)();
  var linesRefs = (0, _react.useRef)(_toConsumableArray(Array(COUNT)).map(function () {
    return (0, _react.createRef)();
  }));

  var _useState = (0, _react.useState)(size),
      _useState2 = _slicedToArray(_useState, 2),
      prevSize = _useState2[0],
      setPrevSize = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      prevDelay = _useState4[0],
      setPrevDelay = _useState4[1];

  var _useState5 = (0, _react.useState)(0),
      _useState6 = _slicedToArray(_useState5, 2),
      prevRepeatDelay = _useState6[0],
      setPrevRepeatDelay = _useState6[1];

  var _useState7 = (0, _react.useState)(0),
      _useState8 = _slicedToArray(_useState7, 2),
      prevRepeat = _useState8[0],
      setPrevRepeat = _useState8[1];

  var center = (0, _react.useMemo)(function () {
    return prevSize / 2;
  }, [prevSize]);
  var circleStrokeWidth = (0, _react.useMemo)(function () {
    return prevSize * CIRCLE_STROKE_WIDTH / 100;
  }, [prevSize]);
  var lineStrokeWidth = (0, _react.useMemo)(function () {
    return prevSize * LINE_STROKE_WIDTH / 100;
  }, [prevSize]);
  var animateCircle = (0, _react.useCallback)(function () {
    var timeline = _gsap["default"].timeline();

    timeline.fromTo(circleRef.current, 0.7, {
      scale: 0,
      transformOrigin: "center"
    }, {
      scale: 1,
      ease: _gsap.Power4.easeOut
    }).fromTo(circleRef.current, 0.4, {
      attr: {
        "stroke-width": circleStrokeWidth
      }
    }, {
      attr: {
        "stroke-width": 0
      },
      ease: _gsap.Power4.easeOut
    }, "-=0.3").fromTo(circleRef.current, 0.3, {
      opacity: 1
    }, {
      opacity: 0,
      ease: _gsap.Power4.easeOut
    }, "-=0.2");
    return timeline;
  }, [circleStrokeWidth]);
  var animateLines = (0, _react.useCallback)(function () {
    var timelines = [];
    var angle = Math.PI / (COUNT / 4);
    var radius = prevSize * LINES_RADIUS / 100;
    linesRefs.current.forEach(function (ref, i) {
      var timeline = _gsap["default"].timeline({
        delay: Math.floor(i / (COUNT / 2)) * 0.4
      });

      var x = center + radius * Math.cos(i * angle);
      var y = center + radius * Math.sin(i * angle);
      var start = {
        x2: x,
        y2: y
      };
      var end = {
        x1: x,
        y1: y
      };
      timeline.fromTo(ref.current, 1, {
        attr: {
          x2: center,
          y2: center
        }
      }, {
        attr: start,
        ease: _gsap.Power4.easeOut
      }).fromTo(ref.current, 1, {
        attr: {
          x1: center,
          y1: center
        }
      }, {
        attr: end,
        ease: _gsap.Power4.easeOut
      }, "-=0.9");
      timelines.push(timeline);
    });
    return timelines;
  }, [center, prevSize]);
  var explode = (0, _react.useCallback)(function () {
    var linesTimelines = animateLines();
    var circletimeline = animateCircle();
    TIME_LINE = _gsap["default"].timeline({
      delay: prevDelay,
      repeat: prevRepeat,
      repeatDelay: prevRepeatDelay,
      onComplete: onComplete,
      onStart: onStart,
      onRepeat: onRepeat
    });
    TIME_LINE.add(linesTimelines, 0);
    TIME_LINE.add(circletimeline, 0.3);
  }, [prevRepeat, prevDelay, prevRepeatDelay, onComplete, onStart, onRepeat, animateCircle, animateLines]);
  (0, _react.useEffect)(function () {
    setPrevSize(size);
    setPrevDelay(delay);
    setPrevRepeatDelay(repeatDelay);
    setPrevRepeat(repeat);
  }, [size, delay, repeatDelay, repeat]);
  (0, _react.useEffect)(function () {
    if (TIME_LINE) TIME_LINE.kill();
    explode();
  }, [explode]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: className,
    style: _objectSpread({
      width: prevSize,
      height: prevSize,
      position: "relative"
    }, style)
  }, /*#__PURE__*/_react["default"].createElement("svg", {
    width: prevSize,
    height: prevSize
  }, linesRefs.current.map(function (ref, i) {
    return /*#__PURE__*/_react["default"].createElement("line", {
      key: i,
      x1: center,
      y1: center,
      x2: center,
      y2: center,
      ref: ref,
      strokeWidth: lineStrokeWidth,
      stroke: color
    });
  })), /*#__PURE__*/_react["default"].createElement(_Circle["default"], {
    width: prevSize,
    height: prevSize,
    r: "47%",
    strokeWidth: circleStrokeWidth,
    color: color,
    shapeRef: circleRef,
    style: {
      position: "absolute",
      left: 0,
      top: 0
    }
  }));
}
