// react-explode/Negros.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Negros;

var _react = _interopRequireWildcard(require("react"));

var _gsap = _interopRequireWildcard(require("gsap"));

var _SineWave = _interopRequireWildcard(require("./Icons/SineWave"));

var _Circle = _interopRequireDefault(require("./Icons/Circle"));

var _Stick = _interopRequireWildcard(require("./Icons/Stick"));

var _useGSAPAnimateStroke = _interopRequireDefault(require("./hooks/useGSAPAnimateStroke"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

// animation configs:
var TIME_LINE;
var SINE_WAVE_LENGTH = 25;
var SINE_WAVE_WIDTHS = [35, 42.5];
var SINE_WAVE_HEIGHT = 3.75;
var INIT_RADIUS = 6;
var STICK_WIDTH = 8.75;
var STICK_HEIGHT = 15;
var STICK_RADIUS = 12.5;
var STROKE_WIDTH = 0.5;
var LINE_STROKE_WIDTH = 1.5;
var SINE_WAVE_STROKE_WIDTHS = [1, 2];
var LINE_X = 2;

function Negros(_ref) {
  var size = _ref.size,
      delay = _ref.delay,
      repeatDelay = _ref.repeatDelay,
      repeat = _ref.repeat,
      style = _ref.style,
      onComplete = _ref.onComplete,
      onStart = _ref.onStart,
      onRepeat = _ref.onRepeat,
      className = _ref.className;
  var lineRef = (0, _react.useRef)();
  var circlesRefs = (0, _react.useRef)(_toConsumableArray(Array(1)).map(function () {
    return (0, _react.createRef)();
  }));
  var sinewavesRefs = (0, _react.useRef)(_toConsumableArray(Array(8)).map(function () {
    return (0, _react.createRef)();
  }));
  var sticksRefs = (0, _react.useRef)(_toConsumableArray(Array(9)).map(function () {
    return {
      svg: (0, _react.createRef)(),
      shape: (0, _react.createRef)()
    };
  }));

  var _useState = (0, _react.useState)(size),
      _useState2 = _slicedToArray(_useState, 2),
      prevSize = _useState2[0],
      setPrevSize = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      prevDelay = _useState4[0],
      setPrevDelay = _useState4[1];

  var _useState5 = (0, _react.useState)(0),
      _useState6 = _slicedToArray(_useState5, 2),
      prevRepeatDelay = _useState6[0],
      setPrevRepeatDelay = _useState6[1];

  var _useState7 = (0, _react.useState)(0),
      _useState8 = _slicedToArray(_useState7, 2),
      prevRepeat = _useState8[0],
      setPrevRepeat = _useState8[1];

  var angle = Math.PI / 2;
  var stickAngle = 2 * Math.PI / 9;
  var prefixAngle = Math.PI / 4;
  var lineStrokeWidth = (0, _react.useMemo)(function () {
    return LINE_STROKE_WIDTH * prevSize / 100;
  }, [prevSize]);
  var animateLine = (0, _react.useCallback)(function () {
    var timeline = _gsap["default"].timeline({
      repeat: 1,
      yoyo: true
    });

    var lineX = LINE_X * prevSize / 100;
    timeline.set(lineRef.current, {
      attr: {
        "stroke-width": 0
      }
    });
    timeline.fromTo(lineRef.current, 0.2, {
      attr: {
        "stroke-width": 0
      }
    }, {
      attr: {
        "stroke-width": lineStrokeWidth
      }
    });
    timeline.fromTo(lineRef.current, 0.3, {
      attr: {
        x2: lineX
      }
    }, {
      attr: {
        x2: prevSize - lineX
      },
      ease: _gsap.Power4.easeIn
    }, "<");
    timeline.fromTo(lineRef.current, 0.3, {
      attr: {
        x1: lineX
      }
    }, {
      attr: {
        x1: prevSize - lineX
      },
      ease: _gsap.Power4.easeOut
    });
    timeline.to(lineRef.current, 0.3, {
      attr: {
        "stroke-width": 0
      },
      ease: _gsap.Power4.easeInOut
    }, "-=0.3");
    return timeline;
  }, [prevSize, lineStrokeWidth]);
  var animateSticks = (0, _react.useCallback)(function () {
    var timelines = [];
    var width = STICK_WIDTH / 2 * prevSize / 100;
    var height = STICK_HEIGHT / 2 * prevSize / 100;
    var radius = STICK_RADIUS * prevSize / 100;
    sticksRefs.current.forEach(function (_ref2, i) {
      var svg = _ref2.svg,
          shape = _ref2.shape;

      var timeline = _gsap["default"].timeline();

      var x = Math.cos(i * stickAngle);
      var y = Math.sin(i * stickAngle);
      timeline.set(shape.current, {
        attr: {
          "stroke-dasharray": "0 ".concat(_Stick.length),
          "stroke-dashoffset": 0,
          "stroke-width": 0
        }
      });
      timeline.fromTo(svg.current, 1, {
        x: -width,
        y: -height
      }, {
        x: radius * x - width,
        y: radius * y - height,
        ease: _gsap.Power4.easeInOut
      });
      timeline.to(shape.current, 0.2, {
        attr: {
          "stroke-width": 1
        }
      }, "<");
      timeline.to(shape.current, {
        keyframes: [{
          attr: {
            "stroke-dasharray": "30 ".concat(_Stick.length - 30),
            "stroke-dashoffset": -20
          },
          duration: 0.5
        }, {
          attr: {
            "stroke-dasharray": "0 ".concat(_Stick.length),
            "stroke-dashoffset": _Stick.length * -1
          },
          duration: 0.7
        }],
        ease: _gsap.Power4.easeInOut
      }, "-=0.2");
      timeline.to(shape.current, 0.3, {
        attr: {
          "stroke-width": 0
        },
        ease: _gsap.Power4.easeInOut
      }, "-=0.3");
      timelines.push(timeline);
    }, []);
    return timelines;
  }, [prevSize, stickAngle]);
  var animateCircles = (0, _react.useCallback)(function () {
    var timelines = [];
    var strokeWidth = prevSize * STROKE_WIDTH / 100;
    circlesRefs.current.forEach(function (ref) {
      var timeline = _gsap["default"].timeline({
        repeat: 1,
        repeatDelay: 0
      });

      timeline.set(ref.current, {
        attr: {
          "stroke-width": 15,
          opacity: 1
        }
      });
      timeline.fromTo(ref.current, 0.7, {
        attr: {
          r: 0,
          "stroke-width": 15
        }
      }, {
        attr: {
          r: "45%",
          "stroke-width": strokeWidth
        },
        ease: _gsap.Power4.easeOut
      });
      timeline.to(ref.current, 0.7, {
        attr: {
          "stroke-width": 0
        },
        ease: _gsap.Power4.easeOut
      }, "-=0.7");
      timeline.fromTo(ref.current, 0.7, {
        opacity: 1
      }, {
        opacity: 0
      }, "-=0.6");
      timelines.push(timeline);
    });
    return timelines;
  }, [prevSize]);
  var animateSineWaveStroke = (0, _useGSAPAnimateStroke["default"])({
    length: 100,
    totalLength: _SineWave.length,
    speed: 1.3
  });
  var animateSinewaves = (0, _react.useCallback)(function () {
    var timelines = [];
    sinewavesRefs.current.forEach(function (ref, i) {
      var timeline = _gsap["default"].timeline({
        delay: 0.05 * (i % 2)
      });

      animateSineWaveStroke({
        elem: ref.current,
        timeline: timeline,
        strokeWidth: SINE_WAVE_STROKE_WIDTHS[i % 2]
      });
      timelines.push(timeline);
    });
    return timelines;
  }, [animateSineWaveStroke]);
  var explode = (0, _react.useCallback)(function () {
    var lineTimeline = animateLine();
    var sticksTimelines = animateSticks();
    var circlesTimelines = animateCircles();
    var sinewavesTimelines = animateSinewaves();
    TIME_LINE = _gsap["default"].timeline({
      repeat: prevRepeat,
      delay: prevDelay,
      repeatDelay: prevRepeatDelay,
      onStart: onStart,
      onComplete: onComplete,
      onRepeat: onRepeat
    });
    TIME_LINE.add(lineTimeline, 0);
    TIME_LINE.add(sinewavesTimelines, "-=1.3");
    TIME_LINE.add(circlesTimelines, 0.2);
    TIME_LINE.add(sticksTimelines, "-=2.1");
  }, [prevRepeat, prevDelay, prevRepeatDelay, onComplete, onStart, onRepeat, animateLine, animateSticks, animateCircles, animateSinewaves]);
  (0, _react.useEffect)(function () {
    if (TIME_LINE) TIME_LINE.kill();
    explode();
  }, [explode]);
  (0, _react.useEffect)(function () {
    setPrevSize(size);
    setPrevDelay(delay);
    setPrevRepeatDelay(repeatDelay);
    setPrevRepeat(repeat);
  }, [size, delay, repeatDelay, repeat]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: className,
    style: _objectSpread({
      width: prevSize,
      height: prevSize,
      position: "relative"
    }, style)
  }, circlesRefs.current.map(function (ref, i) {
    return /*#__PURE__*/_react["default"].createElement(_Circle["default"], {
      key: i,
      shapeRef: ref,
      width: "35%",
      height: "35%",
      color: "#fe4a49",
      style: {
        position: "absolute",
        transform: "translate(-50%, -50%)",
        top: "50%",
        left: "50%"
      }
    });
  }), Array.from(Array(4)).map(function (_, i) {
    return Array.from(Array(2)).map(function (_, j) {
      return /*#__PURE__*/_react["default"].createElement(_SineWave["default"], {
        key: j + 2 * i,
        width: "".concat(SINE_WAVE_WIDTHS[j % 2], "%"),
        height: "".concat(SINE_WAVE_HEIGHT, "%"),
        strokeWidth: 0,
        shapeRef: sinewavesRefs.current[j + 2 * i],
        dasharray: "0 ".concat(_SineWave.length),
        color: "#2ab7ca",
        style: {
          position: "absolute",
          left: "".concat(50 + INIT_RADIUS * Math.cos(i * angle - prefixAngle), "%"),
          top: "".concat(50 + INIT_RADIUS * Math.sin(i * angle - prefixAngle) + j * 2.5, "%"),
          transform: "\n                  rotate(".concat(Math.floor(i / 2) * 180, "deg)"),
          transformOrigin: "0 0"
        }
      });
    });
  }), sticksRefs.current.map(function (_ref3, i) {
    var svg = _ref3.svg,
        shape = _ref3.shape;
    return /*#__PURE__*/_react["default"].createElement(_Stick["default"], {
      key: i,
      ref: svg,
      shapeRef: shape,
      width: "".concat(STICK_WIDTH, "%"),
      height: "".concat(STICK_HEIGHT, "%"),
      dasharray: "0 ".concat(_Stick.length),
      color: "#fe4a49",
      style: {
        position: "absolute",
        left: "50%",
        top: "50%",
        transform: "\n              translate(-50%, -50%)\n              rotate(".concat(40 * i + 110, "deg)\n            "),
        transformOrigin: "center"
      }
    });
  }), /*#__PURE__*/_react["default"].createElement("svg", {
    width: prevSize,
    height: prevSize
  }, /*#__PURE__*/_react["default"].createElement("line", {
    ref: lineRef,
    x1: "".concat(LINE_X, "%"),
    y1: "50%",
    x2: "".concat(LINE_X, "%"),
    y2: "50%",
    stroke: "#fed766",
    strokeLinecap: "round",
    strokeWidth: 0
  })));
}
