// react-explode/Palawan.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Palawan;

var _react = _interopRequireWildcard(require("react"));

var _gsap = _interopRequireWildcard(require("gsap"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var SQUARE = [];
var STROKE_WIDTH = 2.5;
var TIME_LINE = null;

function Palawan(_ref) {
  var size = _ref.size,
      delay = _ref.delay,
      repeatDelay = _ref.repeatDelay,
      repeat = _ref.repeat,
      style = _ref.style,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? "white" : _ref$color,
      onComplete = _ref.onComplete,
      onStart = _ref.onStart,
      onRepeat = _ref.onRepeat,
      className = _ref.className;

  var _useState = (0, _react.useState)(size),
      _useState2 = _slicedToArray(_useState, 2),
      prevSize = _useState2[0],
      setPrevSize = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      prevDelay = _useState4[0],
      setPrevDelay = _useState4[1];

  var _useState5 = (0, _react.useState)(0),
      _useState6 = _slicedToArray(_useState5, 2),
      prevRepeatDelay = _useState6[0],
      setPrevRepeatDelay = _useState6[1];

  var _useState7 = (0, _react.useState)(0),
      _useState8 = _slicedToArray(_useState7, 2),
      prevRepeat = _useState8[0],
      setPrevRepeat = _useState8[1];

  var center = size / 2;
  var sizes = [size * 0.58, size * 0.3, size * 0.3, size * 0.1, size * 0.1];
  var halfs = [sizes[0] / 2, sizes[1] / 2, sizes[3] / 2];
  var front = center - halfs[0];
  var back = center + halfs[0];
  var positions = [{
    x: center,
    y: center
  }, {
    x: center,
    y: back
  }, {
    x: center,
    y: front
  }, {
    x: front,
    y: front
  }, {
    x: back,
    y: back
  }];
  var strokeWidth = Math.ceil(prevSize * STROKE_WIDTH / 100);
  var explode = (0, _react.useCallback)(function () {
    var ease = _gsap.Power4.easeOut;
    var center = prevSize / 2;
    var sizes = [prevSize * 0.58, prevSize * 0.3, prevSize * 0.3, prevSize * 0.1, prevSize * 0.1];
    var halfs = [sizes[0] / 2, sizes[1] / 2, sizes[3] / 2];
    var front = center - halfs[0];
    var back = center + halfs[0];
    var positions = {
      start: [{
        x: center,
        y: center
      }, {
        x: center,
        y: back
      }, {
        x: center,
        y: front
      }, {
        x: front,
        y: front
      }, {
        x: back,
        y: back
      }],
      end: [{
        x: front,
        y: front
      }, {
        x: center - halfs[1],
        y: back - halfs[1]
      }, {
        x: center - halfs[1],
        y: front - halfs[1]
      }, {
        x: front - halfs[2],
        y: front - halfs[2]
      }, {
        x: back - halfs[2],
        y: back - halfs[2]
      }]
    };
    var timelines = [];
    var delays = [0, 0.3, 0.3, 0.5, 0.5];

    for (var i = 0; i < SQUARE.length; i++) {
      var square = SQUARE[i];
      var _size = sizes[i];
      var start = positions.start[i];
      var end = positions.end[i];
      var _delay = delays[i];

      var timeline = _gsap["default"].timeline({
        delay: _delay,
        onStart: i === 0 && onStart,
        onComplete: i === 0 && onComplete,
        onRepeat: i === 0 && onRepeat
      });

      timeline.fromTo(square, 1, {
        attr: {
          x: start.x,
          y: start.y,
          width: 0,
          height: 0
        },
        ease: ease
      }, {
        attr: {
          x: end.x,
          y: end.y,
          width: _size,
          height: _size
        },
        ease: ease
      }).fromTo(square, 1, {
        attr: {
          "stroke-width": strokeWidth
        },
        ease: ease
      }, {
        attr: {
          "stroke-width": 0
        },
        ease: ease
      }, "-=0.9");
      timelines.push(timeline);
    }

    TIME_LINE = _gsap["default"].timeline({
      delay: prevDelay,
      repeat: prevRepeat,
      repeatDelay: prevRepeatDelay
    });
    TIME_LINE.add(timelines);
  }, [prevSize, prevRepeat, prevDelay, prevRepeatDelay, onComplete, onStart, onRepeat, strokeWidth]);
  (0, _react.useEffect)(function () {
    setPrevSize(size);
    setPrevDelay(delay);
    setPrevRepeatDelay(repeatDelay);
    setPrevRepeat(repeat);
  }, [size, delay, repeatDelay, repeat]);
  (0, _react.useEffect)(function () {
    explode();
  }, [explode]);
  (0, _react.useEffect)(function () {
    TIME_LINE.kill();
    explode();
  });
  return /*#__PURE__*/_react["default"].createElement("svg", {
    width: prevSize,
    height: prevSize,
    className: className,
    style: _objectSpread({}, style, {
      position: "relative"
    })
  }, _toConsumableArray(Array(5)).map(function (_, i) {
    var _positions$i = positions[i],
        x = _positions$i.x,
        y = _positions$i.y;
    return /*#__PURE__*/_react["default"].createElement("rect", {
      key: i,
      x: x,
      y: y,
      width: 0,
      height: 0,
      stroke: color,
      strokeWidth: strokeWidth,
      fill: "none",
      ref: function ref(el) {
        return SQUARE[i] = el;
      },
      style: {
        transform: "rotate(-50deg)",
        transformOrigin: "".concat(center, "px ").concat(center, "px")
      }
    });
  }));
}
