// react-explode/Polillo.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Polillo;

var _react = _interopRequireWildcard(require("react"));

var _gsap = _interopRequireWildcard(require("gsap"));

var _ZigZag = _interopRequireDefault(require("./Icons/ZigZag"));

var _Cross = _interopRequireDefault(require("./Icons/Cross"));

var _Circle = _interopRequireDefault(require("./Icons/Circle"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var RADIUSES = [47, 30];
var CROSS_SIZE = 17;
var ZIGZAG_WIDTH = 10;
var ZIGZAG_HEIGHT = 20;
var STROKE_WIDTH = 0.1;
var TIME_LINE = null;

function Polillo(_ref) {
  var _ref$size = _ref.size,
      size = _ref$size === void 0 ? 400 : _ref$size,
      delay = _ref.delay,
      repeatDelay = _ref.repeatDelay,
      repeat = _ref.repeat,
      style = _ref.style,
      onComplete = _ref.onComplete,
      onStart = _ref.onStart,
      onRepeat = _ref.onRepeat,
      className = _ref.className;
  var zigzagRefs = (0, _react.useRef)(_toConsumableArray(Array(9)).map(function () {
    return (0, _react.createRef)();
  }));
  var crossRefs = (0, _react.useRef)(_toConsumableArray(Array(9)).map(function () {
    return (0, _react.createRef)();
  }));
  var circleRefs = (0, _react.useRef)(_toConsumableArray(Array(2)).map(function () {
    return (0, _react.createRef)();
  }));

  var _useState = (0, _react.useState)(size),
      _useState2 = _slicedToArray(_useState, 2),
      prevSize = _useState2[0],
      setPrevSize = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      prevDelay = _useState4[0],
      setPrevDelay = _useState4[1];

  var _useState5 = (0, _react.useState)(0),
      _useState6 = _slicedToArray(_useState5, 2),
      prevRepeatDelay = _useState6[0],
      setPrevRepeatDelay = _useState6[1];

  var _useState7 = (0, _react.useState)(0),
      _useState8 = _slicedToArray(_useState7, 2),
      prevRepeat = _useState8[0],
      setPrevRepeat = _useState8[1];

  var strokeWidth = prevSize * STROKE_WIDTH / 100;
  var radiuses = (0, _react.useMemo)(function () {
    return RADIUSES.map(function (radius) {
      return prevSize * radius / 100;
    });
  }, [prevSize]);
  var crossSize = (0, _react.useMemo)(function () {
    return CROSS_SIZE / 2 * prevSize / 100;
  }, [prevSize]);
  var angle = 2 * (Math.PI / 9);
  var animateZigzags = (0, _react.useCallback)(function () {
    var timelines = [];
    zigzagRefs.current.forEach(function (ref, i) {
      var timeline = _gsap["default"].timeline();

      var cos = Math.cos(2 * (Math.PI / 9) * i);
      var sin = Math.sin(2 * (Math.PI / 9) * i);
      var x = ZIGZAG_WIDTH / 2 * prevSize / 100;
      var y = ZIGZAG_HEIGHT / 2 * prevSize / 100;
      timeline.fromTo(ref.current, 1, {
        x: -x,
        y: -y
      }, {
        x: "".concat(-x + radiuses[0] * cos),
        y: "".concat(-y + radiuses[0] * sin),
        ease: _gsap.Power4.easeOut
      });
      timeline.fromTo(ref.current, 2, {
        rotation: 0,
        scale: 1
      }, {
        rotation: 360,
        scale: 0,
        transformOrigin: "50% 50%",
        ease: _gsap.Power4.easeOut
      }, 0);
      timelines.push(timeline);
    }, []);
    return timelines;
  }, [prevSize, radiuses]);
  var animateCrosses = (0, _react.useCallback)(function () {
    var timelines = [];
    crossRefs.current.forEach(function (ref, i) {
      var timeline = _gsap["default"].timeline();

      var cos = Math.cos(angle * i);
      var sin = Math.sin(angle * i);
      timeline.fromTo(ref.current, 1, {
        x: -crossSize,
        y: -crossSize
      }, {
        x: -crossSize + radiuses[1] * cos,
        y: -crossSize + radiuses[1] * sin,
        ease: _gsap.Power4.easeOut
      }, "<");
      timeline.fromTo(ref.current, 2, {
        rotation: 0,
        scale: 1
      }, {
        rotation: 360,
        scale: 0,
        transformOrigin: "50% 50%",
        ease: _gsap.Power4.easeOut
      }, "<");
      timelines.push(timeline);
    });
    return timelines;
  }, [radiuses, crossSize, angle]);
  var animateCircles = (0, _react.useCallback)(function () {
    var timelines = [];
    circleRefs.current.forEach(function (ref, i) {
      var timeline = _gsap["default"].timeline({
        delay: 0.3 * i
      });

      timeline.set(ref.current, {
        attr: {
          "stroke-width": 1
        }
      });
      timeline.fromTo(ref.current, 1, {
        scale: 0,
        transformOrigin: "center"
      }, {
        scale: 1,
        ease: _gsap.Power4.easeOut
      });
      timeline.fromTo(ref.current, 0.3, {
        attr: {
          "stroke-width": 1
        },
        transformOrigin: "center"
      }, {
        attr: {
          "stroke-width": 0
        },
        ease: _gsap.Power4.easeOut
      }, "-=0.6");
      timeline.fromTo(ref.current, 1, {
        opacity: 1
      }, {
        opacity: 0,
        ease: _gsap.Power4.easeOut
      }, "-=0.3");
      timelines.push(timeline);
    });
    return timelines;
  }, []);
  var explode = (0, _react.useCallback)(function () {
    var zigzagsTimelines = animateZigzags();
    var crossesTimelines = animateCrosses();
    var circlesTimelines = animateCircles();
    TIME_LINE = _gsap["default"].timeline({
      repeat: prevRepeat,
      delay: prevDelay,
      repeatDelay: prevRepeatDelay,
      onStart: onStart,
      onComplete: onComplete,
      onRepeat: onRepeat
    });
    TIME_LINE.add(zigzagsTimelines, 0);
    TIME_LINE.add(crossesTimelines, 0);
    TIME_LINE.add(circlesTimelines, 0.1);
  }, [onComplete, onStart, onRepeat, prevDelay, prevRepeat, prevRepeatDelay, animateZigzags, animateCrosses, animateCircles]);
  (0, _react.useEffect)(function () {
    if (TIME_LINE) TIME_LINE.kill();
    explode();
  }, [explode]);
  (0, _react.useEffect)(function () {
    setPrevSize(size);
    setPrevDelay(delay);
    setPrevRepeatDelay(repeatDelay);
    setPrevRepeat(repeat);
  }, [size, delay, repeatDelay, repeat]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: className,
    style: _objectSpread({
      width: prevSize,
      height: prevSize,
      position: "relative"
    }, style)
  }, zigzagRefs.current.map(function (ref, i) {
    return /*#__PURE__*/_react["default"].createElement(_ZigZag["default"], {
      key: i,
      color: "#fe4a49",
      style: {
        position: "absolute",
        top: "50%",
        left: "50%",
        transform: "translate(-50%, -50%)",
        width: "".concat(ZIGZAG_WIDTH, "%"),
        height: "".concat(ZIGZAG_HEIGHT, "%")
      },
      ref: ref
    });
  }), crossRefs.current.map(function (ref, i) {
    return /*#__PURE__*/_react["default"].createElement(_Cross["default"], {
      key: i,
      ref: ref,
      color: "#2ab7ca",
      style: {
        position: "absolute",
        top: "50%",
        left: "50%",
        transform: "translate(-50%, -50%)",
        width: "".concat(CROSS_SIZE, "%"),
        height: "".concat(CROSS_SIZE, "%")
      }
    });
  }), circleRefs.current.map(function (ref, i) {
    return /*#__PURE__*/_react["default"].createElement(_Circle["default"], {
      key: i,
      radius: "40%",
      fill: "none",
      color: "#fed766",
      strokeWidth: strokeWidth,
      style: {
        position: "absolute",
        width: "100%",
        height: "100%"
      },
      shapeRef: ref
    });
  }));
}
