// react-explode/Siargao.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Siargao;

var _react = _interopRequireWildcard(require("react"));

var _gsap = _interopRequireWildcard(require("gsap"));

var _Wave = _interopRequireWildcard(require("./Icons/Wave2"));

var _Circle = _interopRequireDefault(require("./Icons/Circle"));

var _useGSAPAnimateStroke = _interopRequireDefault(require("./hooks/useGSAPAnimateStroke"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var TIME_LINE;
var unicornColors = ["#5B41C1", "#5893E0", "#43C2C2", "#58E55C", "#FCF04E", "#FFA500", "#F25A5F", "#FF6A80"];
var LINES_LENGTH = 8;
var CIRCLES = [{
  pos: [50, 50],
  size: 50,
  delay: 0.35
}, {
  pos: [68, 15],
  size: 10,
  delay: 0.6
}, {
  pos: [65, 20],
  size: 7,
  delay: 0.7
}, {
  pos: [32, 84],
  size: 10,
  delay: 0.6
}, {
  pos: [28, 80],
  size: 7,
  delay: 0.7
}];
var CIRCLES_LENGTH = CIRCLES.length;
var LINE_STROKE_WIDTH = 0.8;
var CIRCLE_STROKE_WIDTH = 2;

function Siargao(_ref) {
  var size = _ref.size,
      delay = _ref.delay,
      repeatDelay = _ref.repeatDelay,
      repeat = _ref.repeat,
      style = _ref.style,
      onComplete = _ref.onComplete,
      onStart = _ref.onStart,
      onRepeat = _ref.onRepeat,
      className = _ref.className;
  var linesRefs = (0, _react.useRef)(_toConsumableArray(Array(LINES_LENGTH)).map(function () {
    return (0, _react.createRef)();
  }));
  var circlesRefs = (0, _react.useRef)(_toConsumableArray(Array(CIRCLES_LENGTH)).map(function () {
    return {
      svg: (0, _react.createRef)(),
      shape: (0, _react.createRef)()
    };
  }));

  var _useState = (0, _react.useState)(size),
      _useState2 = _slicedToArray(_useState, 2),
      prevSize = _useState2[0],
      setPrevSize = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      prevDelay = _useState4[0],
      setPrevDelay = _useState4[1];

  var _useState5 = (0, _react.useState)(0),
      _useState6 = _slicedToArray(_useState5, 2),
      prevRepeatDelay = _useState6[0],
      setPrevRepeatDelay = _useState6[1];

  var _useState7 = (0, _react.useState)(0),
      _useState8 = _slicedToArray(_useState7, 2),
      prevRepeat = _useState8[0],
      setPrevRepeat = _useState8[1];

  var animateStroke = (0, _useGSAPAnimateStroke["default"])({
    length: 30,
    totalLength: _Wave.length,
    speed: 1.3
  });
  var animateLines = (0, _react.useCallback)(function () {
    var timelines = [];
    linesRefs.current.forEach(function (ref, i) {
      var timeline = _gsap["default"].timeline({
        delay: i % 4 * 0.03
      });

      animateStroke({
        elem: ref.current,
        timeline: timeline,
        strokeWidth: LINE_STROKE_WIDTH
      });
      timelines.push(timeline);
    });
    return timelines;
  }, [animateStroke]);
  var animateCircles = (0, _react.useCallback)(function () {
    var timelines = [];
    var strokeWidth = prevSize * CIRCLE_STROKE_WIDTH / 100;
    circlesRefs.current.forEach(function (_ref2, i) {
      var svg = _ref2.svg,
          shape = _ref2.shape;

      var timeline = _gsap["default"].timeline({
        delay: CIRCLES[i].delay
      });

      timeline.set(shape.current, {
        attr: {
          "stroke-width": 15,
          r: 0
        }
      });
      timeline.set(svg.current, {
        opacity: 1
      });
      timeline.fromTo(shape.current, 0.5, {
        attr: {
          r: 0,
          "stroke-width": 15
        }
      }, {
        attr: {
          r: "45%",
          "stroke-width": strokeWidth
        },
        ease: _gsap.Power4.easeOut
      });
      timeline.to(shape.current, 0.5, {
        attr: {
          "stroke-width": 0
        },
        ease: _gsap.Power4.easeOut
      }, "-=0.5");
      timeline.fromTo(svg.current, 0.4, {
        opacity: 1
      }, {
        opacity: 0
      }, "-=0.4");
      timelines.push(timeline);
    });
    return timelines;
  }, [prevSize]);
  var explode = (0, _react.useCallback)(function () {
    var linesTimelines = animateLines();
    var circleTimeline = animateCircles();
    TIME_LINE = _gsap["default"].timeline({
      repeat: prevRepeat,
      delay: prevDelay,
      repeatDelay: prevRepeatDelay,
      onStart: onStart,
      onComplete: onComplete,
      onRepeat: onRepeat
    });
    TIME_LINE.add(linesTimelines, 0);
    TIME_LINE.add(circleTimeline, 0);
  }, [animateCircles, animateLines, onComplete, onStart, onRepeat, prevRepeat, prevDelay, prevRepeatDelay]);
  (0, _react.useEffect)(function () {
    if (TIME_LINE) TIME_LINE.kill();
    explode();
  }, [explode]);
  (0, _react.useEffect)(function () {
    setPrevSize(size);
    setPrevDelay(delay);
    setPrevRepeatDelay(repeatDelay);
    setPrevRepeat(repeat);
  }, [size, delay, repeatDelay, repeat]);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: className,
    style: _objectSpread({
      width: prevSize,
      height: prevSize,
      position: "relative"
    }, style)
  }, linesRefs.current.map(function (ref, i) {
    return /*#__PURE__*/_react["default"].createElement(_Wave["default"], {
      key: i,
      shapeRef: ref,
      width: "68%",
      height: "50%",
      strokeWidth: 0,
      dasharray: "0 ".concat(_Wave.length),
      color: unicornColors[i],
      style: {
        position: "absolute",
        transform: "\n                translate(-50%, -50%)\n                rotate(".concat(Math.floor(i * 2 / LINES_LENGTH) * 180, "deg)\n            "),
        top: "".concat(50 + (Math.floor(i * 2 / (LINES_LENGTH - 1)) ? -1 : 1) * 20.5 - (LINES_LENGTH - 1) * 2.5 / 2 + i * 2.5, "%"),
        left: "".concat(50 + Math.cos(Math.PI * Math.floor(i * 2 / LINES_LENGTH)) * 15, "%")
      }
    });
  }), CIRCLES.map(function (_ref3, i) {
    var pos = _ref3.pos,
        size = _ref3.size;
    return /*#__PURE__*/_react["default"].createElement(_Circle["default"], {
      key: i,
      shapeRef: circlesRefs.current[i].shape,
      ref: circlesRefs.current[i].svg,
      width: "".concat(size, "%"),
      height: "".concat(size, "%"),
      strokeWidth: 15,
      radius: 0,
      color: unicornColors[unicornColors.length - 1 - i],
      style: {
        position: "absolute",
        transform: "translate(-50%, -50%)",
        left: "".concat(pos[0], "%"),
        top: "".concat(pos[1], "%"),
        opacity: 0
      }
    });
  }));
}
