// react-explode/Siquijor.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Siquijor;

var _react = _interopRequireWildcard(require("react"));

var _gsap = _interopRequireWildcard(require("gsap"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var TARGETS = [];
var COUNTS = [7, 32];
var WIDTHS = [1.75, 0.5];
var RADIUS = 47.5;
var CIRCLE_STROKE = 1;
var CIRCLES = [{
  el: null,
  pos: [50, 50],
  radius: 25
}, {
  el: null,
  pos: [68, 46],
  radius: 40
}, {
  el: null,
  pos: [30, 60],
  radius: 32.5
}];
var TIME_LINE = null;

function Siquijor(_ref) {
  var size = _ref.size,
      delay = _ref.delay,
      repeatDelay = _ref.repeatDelay,
      repeat = _ref.repeat,
      style = _ref.style,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? "white" : _ref$color,
      onComplete = _ref.onComplete,
      onStart = _ref.onStart,
      onRepeat = _ref.onRepeat,
      className = _ref.className;

  var _useState = (0, _react.useState)(size),
      _useState2 = _slicedToArray(_useState, 2),
      prevSize = _useState2[0],
      setPrevSize = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      prevDelay = _useState4[0],
      setPrevDelay = _useState4[1];

  var _useState5 = (0, _react.useState)(0),
      _useState6 = _slicedToArray(_useState5, 2),
      prevRepeatDelay = _useState6[0],
      setPrevRepeatDelay = _useState6[1];

  var _useState7 = (0, _react.useState)(0),
      _useState8 = _slicedToArray(_useState7, 2),
      prevRepeat = _useState8[0],
      setPrevRepeat = _useState8[1];

  var circleStroke = Math.ceil(prevSize * CIRCLE_STROKE / 100);
  var center = prevSize / 2;
  var explode = (0, _react.useCallback)(function () {
    var ease = _gsap.Power4.easeOut;
    var radius = prevSize * RADIUS / 100;
    var timelines = [];
    TIME_LINE = _gsap["default"].timeline({
      repeat: prevRepeat,
      repeatDelay: prevRepeatDelay,
      delay: prevDelay
    });

    for (var i = 0; i < 2; i++) {
      var angle = Math.PI / (COUNTS[i] / 2);

      for (var j = 0; j < COUNTS[i]; j++) {
        var isLast = j >= COUNTS[i] - 1;

        var timeline = _gsap["default"].timeline({
          delay: i >= 1 ? 0 : 0.15,
          onComplete: onComplete && isLast && onComplete.bind(null, i),
          onStart: onStart && isLast && onStart.bind(null, i),
          onRepeat: onRepeat && isLast && onRepeat.bind(null, i)
        });

        var x = center + radius * Math.cos(j * angle);
        var y = center + radius * Math.sin(j * angle);
        var target = TARGETS[i][j];
        var start = {
          x2: x,
          y2: y
        };
        var end = {
          x1: x,
          y1: y
        };
        timeline.fromTo(target, 1, {
          attr: {
            x2: center,
            y2: center
          }
        }, {
          attr: start,
          ease: ease
        }).fromTo(target, 1, {
          attr: {
            x1: center,
            y1: center
          }
        }, {
          attr: end,
          ease: ease
        }, "-=0.9");
        timelines.push(timeline);
      }

      for (var _i2 = 0; _i2 < CIRCLES.length; _i2++) {
        var circle = CIRCLES[_i2];
        var el = circle.el;

        var _radius = prevSize * circle.radius / 100;

        var _timeline = _gsap["default"].timeline({
          delay: (_i2 + 1) * 0.2,
          onComplete: onComplete && onComplete.bind(null, 2 + _i2),
          onStart: onStart && onStart.bind(null, 2 + _i2),
          onRepeat: onRepeat && onRepeat.bind(null, 2 + 1)
        });

        _timeline.fromTo(el, 1, {
          attr: {
            r: 0
          }
        }, {
          attr: {
            r: _radius / 2
          },
          ease: ease
        }).fromTo(el, 1, {
          attr: {
            "stroke-width": circleStroke
          }
        }, {
          attr: {
            "stroke-width": 0
          },
          ease: ease
        }, "-=0.9").fromTo(el, 0.6, {
          attr: {
            opacity: 1
          }
        }, {
          attr: {
            opacity: 0
          },
          ease: ease
        }, "-=0.5");

        timelines.push(_timeline);
      }

      TIME_LINE.add(timelines, 0);
    }
  }, [prevSize, prevRepeat, prevDelay, prevRepeatDelay, onComplete, onStart, onRepeat, center, circleStroke]);
  (0, _react.useEffect)(function () {
    setPrevSize(size);
    setPrevDelay(delay);
    setPrevRepeatDelay(repeatDelay);
    setPrevRepeat(repeat);
  }, [size, delay, repeatDelay, repeat]);
  (0, _react.useEffect)(function () {
    if (TIME_LINE) TIME_LINE.kill();
    explode();
  }, [explode]);
  return /*#__PURE__*/_react["default"].createElement("svg", {
    style: _objectSpread({}, style, {
      position: "relative"
    }),
    width: prevSize,
    height: prevSize,
    className: className
  }, /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, _toConsumableArray(Array(2)).map(function (_, i) {
    var width = Math.ceil(prevSize * WIDTHS[i] / 100);
    TARGETS[i] = [];
    return /*#__PURE__*/_react["default"].createElement(_react.Fragment, {
      key: i
    }, _toConsumableArray(Array(COUNTS[i])).map(function (_, j) {
      return /*#__PURE__*/_react["default"].createElement("line", {
        x1: center,
        y1: center,
        x2: center,
        y2: center,
        ref: function ref(el) {
          return TARGETS[i][j] = el;
        },
        key: j,
        strokeWidth: width,
        stroke: color
      });
    }));
  })), /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, _toConsumableArray(Array(3)).map(function (_, i) {
    var position = CIRCLES[i].pos;
    var x = prevSize * position[0] / 100;
    var y = prevSize * position[1] / 100;
    return /*#__PURE__*/_react["default"].createElement("circle", {
      key: i,
      cx: x,
      cy: y,
      r: 0,
      strokeWidth: circleStroke,
      fill: "none",
      stroke: color,
      ref: function ref(el) {
        return CIRCLES[i].el = el;
      }
    });
  })));
}
